src/openforms/registrations/contrib/objects_api/config.py

```diff
@@ -92,3 +142,46 @@ class ObjectsAPIOptionsSerializer(JsonSchemaSerializerMixin, serializers.Seriali
         ],
         required=False,
     )
+
+    # V2 only fields:
+    variables_mapping = ObjecttypeVariableMappingSerializer(
+        label=_("variables mapping"),
+        many=True,
+        required=False,
+    )
+
+    def validate(self, attrs: dict[str, Any]) -> dict[str, Any]:
+        v1_only_fields = {"content_json", "payment_status_update_json"}
+        v2_only_fields = {"variables_mapping"}
+
+        version = get_from_serializer_data_or_instance("version", attrs, self)
+
+        match version:
+            case VersionChoices.V1:
+                v1_forbidden_fields = v2_only_fields.intersection(attrs)
+                if v1_forbidden_fields:
+                    raise ValidationError(
+                        {
+                            k: _(
+                                "{field_name} shouldn't be provided when version is 1"
+                            ).format(field_name=k)
+                            for k in v1_forbidden_fields
+                        }
+                    )
+            case VersionChoices.V2:
+                v2_forbidden_fields = v1_only_fields.intersection(attrs)
+                if v2_forbidden_fields:
+                    raise ValidationError(
+                        {
+                            k: _(
+                                "{field_name} shouldn't be provided when version is 2"
+                            ).format(field_name=k)
+                            for k in v2_forbidden_fields
+                        }
+                    )
+            case _:
+                raise ValidationError(
+                    {"version": _("Unknown version: {version}").format(version=version)}
+                )
+
+        return attrs
```

Comment:
happy flow not tested - are you planning to write more API endpoint tests or serializer unit tests?

I reckon these paths will be tested in the actual registration flows themselves since they also call `serializer.is_valid()` in there, so I leave it up to you.